### ISSUE TITLE
Batched retrieve of pending (mempool) transactions

### ISSUE THREAD

**Issue by catenocrypt** [trezor/blockbook#397]
Blockbook keeps track of pending (unconfirmed, mempool) transactions, and it returns pending transactions as well when transactions of an address are retrieved.  However, pending transactions cannot be retrieved in a batch, independently of any address.

*Motivation:* In usage scenarios where a high number of addresses are followed, it is not efficient to poll for the transactions of each address.  Instead, it would be useful to be able to retrieve the set of pending transactions (similar how transactions of one block are returned), and then they can be filtered by addresses.

*Proposal:* an API to return current pending transaction, filtered by a timestamp range.
E.g.   `GET /api/v1/mempool/<start-time>`

Use case with periodical retrieval of blocks and pending transactions:
1. Retrieve blockbook status
1. If there is no new block (since last), and no new mempool tx (since last), sleep for some time, start over (Note: current `/api/` returns latest mempool tx timestamp `lastMempoolTime`).
1. If there is new block (or more), retrieve new block(s), process transactions, start over
1. If there is new block, but there are new pending transactions, retrieve new pending transactions (using new API), process them, start over.

**Comment by martinboehm** [comment]
While it would be possible to create such API, I would suggest to use the websocket interface, where you can subscribe to be notified about new blocks and mempool transactions affecting selected addressed.

**Comment by catenocrypt** [comment]
Is it possible to subscribe using websocket for any transactions, or only for those related to a particular address?  Because if it is only per-address, it is not practical to have 10K+ websocket connections for the 10K+ addresses to be watched.

**Comment by martinboehm** [comment]
You do not need 10K+ websocket connections. Create exactly one connection and observe 10K+ addresses. The websocket api `subscribeAddresses` accepts multiple addresses as comma separated value list. 
Blockbook keeps one subscription per connection, it sends notifications about addresses subscribed in the last call to `subscribeAddresses.

**Comment by vikmeup** [comment]
@martinboehm or even better just for subscribe to all incoming mempool transactions and then the client will filter out transactions/addresses he does not need. Having options for both would be great. 

Current use-case we explore on @trustwallet to notify users for incoming transfers from mempool, observing for millions of addresses might be tricky, for option for all addresses might work better in this case.

**Comment by catenocrypt** [comment]
One connection watching 10K+ connections is fine, but at this number the updating of watched address list is in itself an issue (new addresses coming very often).

**Comment by catenocrypt** [comment]
Here is the detailed proposal for the new API.  I will start the implementation and create a PR.

Proposed mempool API
`GET /api/v2/mempool/<fromTime>`

- returns the transactions currently in mempool (not yet in a block), newer than a threshold
- API modeled on /api/v2/block
- filter parameter is a minimum time, transactions older than this are not included (by first seen time).  This way clients can retrieve lices of the mempool in their own pace, periodically, without retrieving transactions multiple times (note: the mempoool can easily contain transactions corresponding to multiple blocks, and there can be transactions staying in the mempool for relatively long).
- if fromTime is missing or 0, all mempool transactions are returned
- paging is supported as in block API
- transaction details are included (as in block API), not only txIDs
- lastTime is the most recent lastTime of the transactions.  This value should be used as fromTime in the next call.
- mempoolSize is the number of transactions in the mempool (unfiltered)
- previousBlockHash is the hash of the last block (similar to block API)
- height is the height of the last block
- txCount is the number of transactions considered in this requests, i.e, filtered
- for transactions, fields like blockHash, blockHeight, confirmations are not present as they do not apply
- Field blockTime is also missing, but time is present, filled with the first seen time of the transaction.


Sample request:
`https://localhost:9130/api/v2/mempool/1588769940`

Sample response:
```
{
    "page": 1,
    "totalPages": 2,
    "itemsOnPage": 1000,
    "lastTime": 1588769950,
    "mempoolSize": 10495,
    "previousBlockHash": "0000000000000000000d9cc496f5aae75804c2493e830d37867a77d34935a757",
    "height": 629212,
    "txCount": 132,
    "txs": [
        {
            "txid": "aaa04254e7c710ef5e4eeaa0a205402e5f1b07dbdced7b828918f309cd334dc3",
            "vin": [
                {
                    "n": 0,
                    "isAddress": false,
                    "value": "0"
                }
            ],
            "vout": [
                {
                    "value": "1266211215",
                    "n": 0,
                    "addresses": [
                        "147SwRQdpCfj5p8PnfsXV2SsVVpVcz3aPq"
                    ],
                    "isAddress": true
                },
                {
                    "value": "0",
                    "n": 1,
                    "addresses": [
                        "OP_RETURN aa21a9ed015e62c0da386fa8ce141c3e8c8267d1de67a6cba1d74a1fe29b6cba2053bf8c"
                    ],
                    "isAddress": false
                },
                {
                    "value": "0",
                    "n": 2,
                    "addresses": [
                        "OP_RETURN b9e11b6ddff0faf533edc00cb7631e2b81fc3b1f5f19836cb7af7fb6f7a311f583deb5a9"
                    ],
                    "isAddress": false
                }
            ],
            "time": 1588769950,
            "value": "1266211215",
            "valueIn": "0",
            "fees": "0"
        },
        ...
    ]
}
```

**Comment by martinboehm** [comment]
Hi, thank you for the proposal. The problem is that to get the TX details, Blockbook would have to do a request for each transaction to the backend, which is inherently slow (only tens of request per second). It would take many seconds or even minutes to query the whole mempool.

I would still prefer the streaming websocket interface. If the number of addresses to watch is high and changes a lot, maybe a modification of the subscription API, to send all new mempool transactions, would fulfil your needs.

**Comment by catenocrypt** [comment]
I was not aware of the backend boundary, this is indeed problematic (even if time-based filtering is done first, and transaction detail fetch only for the filtered TXs).

I noted though, that worker.GetTransaction() uses txCache internally, doesn't this means that each TX is brought only once from the backend?

How about a batched GetTransactions(txids []string), would that help?

I am fine with the websocket interface extension too, I will check into it.

**Comment by martinboehm** [comment]
We use the tx cache as queries for transactions are slow. However, mempool transactions are not cached at all.

**Comment by catenocrypt** [comment]
Extended websocket subscription would look like this:
- Currently it is possible to subscribe to a list of addresses (subscribeAddresses)
- Similar to this there would by new option to subscribe to *any* address (subscribeAllAddresses)
- A TX notification is sent to clients subscribed to that address, or to any address.
I don't know if it makes sense (or easily doable) to limit this to mempool addresses.
Does this sound acceptable?

**Comment by martinboehm** [comment]
It is implemented in PR #571
